[box] New Integration

djptek opened the pull request:
- Enhancement

## What does this PR do?

Adds integration for BOX events.

## Checklist

- [Y] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [Y] I have verified that all data streams collect metrics or logs.
- [Y] I have added an entry to my package's `changelog.yml` file.
- [Y] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [x] Complete system tests

Add Shield Alerts

- [x] Suspicious locations
- [x] Suspicious sessions
- [x] Anomalous downloads
- [x] Malicious content

- [x] Update Dashboards
- [x] Update Screenshots
- [x] Update docs template `_dev/build/docs/README.md`

## How to test this PR locally

Use `elastic-package test` to run the tests.

The pipeline tests for the events were created by ingesting data directly from the Box API using Filebeat HTTPJSON with no processors applied, then cut-and-paste of the message field for each resultant Elasticsearch Document to the relevant inputs.

An additional test was added using fake events, one corresponding to each of [all documented box event types](https://developer.box.com/reference/resources/event/#param-event_type). This test is intended to exercise the mapping of the box `event_type` field value to the corresponding values of ECS fields `event.category` and `event.type`.

The simulated shield alerts were created by:
- composing the alerts in accordance with the [Box Shield Alerts documentation](https://developer.box.com/guides/events/event-triggers/shield-alert-events/) and publishing these via a web proxy
- ingest of all alerts using Filebeat running 4 distinct configurations, each based on the relevant `httpjson.yml.hbs` for the relevant data stream, to ensure that the distinct `response.split.split` and `drop_event` pipeline processors were applied to denormalise alerts to constituent documents and select these for routing to the relevant data-stream pipeline

Box Shield Alerts should be considered as Beta until such time as access to a suitable target system with alerts relating to suspected malign activity is available.
 
To run an end-to-end test with live data, you will need an account on Box, a [developer account](https://app.box.com/developers/console) is fine, don't need enterprise.

Add some test data to that and copy/delete/upload/download/view the data to create some events.

Have your `box` credentials available
- client.id
- client.secret 
- subject_id

(do not store these in Github)

see the documentation of the `box_events` Integration for details.

Start the stack using `elastic-package stack up -v -d --version 8.3.0`

Add and configure your `box_events` Integration using your Box credentials. 

![image](https://user-images.githubusercontent.com/11043991/188626816-aeb1dcb9-8ffc-4a0e-8cd3-d977e87398d7.png)

Ensure to scroll down to

<img width="784" alt="image" src="https://user-images.githubusercontent.com/11043991/188625149-5864d0fc-2c42-4068-835c-2e557459939c.png">

and assign the policy to `Existing hosts`->  `Agent Policy` -> `Elastic-Agent`

Generate some events in Box by Upload/Download/Delete/Visualisation of files then go to the `[Logs Box Events Integration] Events` dashboard to verify that events were ingested and are visible.

To add Simulated Shield Events, you can cut and paste the contents of [this `_bulk` request](https://gist.githubusercontent.com/djptek/aa74e727a02c3b82d85c6182e615c2d8/raw/b76f1319b4ac5492ebbca2f6028f5e94c61c9954/box_alerts.ndjson) into `Dev Tools` and send the request to the stack. This will ingest a selection of alerts to validate the `[Logs Box Events Integration] Box Shield Alerts` dashboard, you will need to set the time picker to July 2022.

The Simulated Shield Alerts were created by manually appending raw `agent` etc. fields created during the end-to-end tests for each alert's pipeline test document(s), and then rectifying the data-stream for each.

## Related issues

- Relates https://github.com/elastic/integrations/issues/1151

## Screenshots

For Kibana UI options see test procedure, above.

# Dashboards

Showing Elastic Stack 7.17.0 compatible dashboards

## Box Events

![image](https://user-images.githubusercontent.com/11043991/190112744-4887d0c0-bb39-4770-afc8-3fa357b7d040.png)

## Box Shield Alerts

![image](https://user-images.githubusercontent.com/11043991/190112832-28bc0fbc-ee83-472d-a757-00a2a2f8f89b.png)

<img width="1525" alt="image" src="https://user-images.githubusercontent.com/11043991/188699759-72d47bc7-51bd-4510-82e3-3d92c2d42e1a.png">

<img width="1526" alt="image" src="https://user-images.githubusercontent.com/11043991/188699817-ee1dd40a-fceb-49dc-8ad8-03c48cb432f5.png">

# Test Results

```
% elastic-package test                                                                 (box)integrations
Run test suite for the package
Run asset tests for the package
--- Test results for package: box_events - START ---
╭────────────┬─────────────────────────────┬───────────┬─────────────────────────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE    │ DATA STREAM                 │ TEST TYPE │ TEST NAME                                                                   │ RESULT │ TIME ELAPSED │
├────────────┼─────────────────────────────┼───────────┼─────────────────────────────────────────────────────────────────────────────┼────────┼──────────────┤
│ box_events │                             │ asset     │ dashboard box_events-ce6fbf50-2df9-11ed-8003-6d5721603181 is loaded         │ PASS   │      4.062µs │
│ box_events │                             │ asset     │ dashboard box_events-ff3d9940-2e03-11ed-8003-6d5721603181 is loaded         │ PASS   │        110ns │
│ box_events │ anomalous_download_alerts   │ asset     │ index_template logs-box_events.anomalous_download_alerts is loaded          │ PASS   │        276ns │
│ box_events │ anomalous_download_alerts   │ asset     │ ingest_pipeline logs-box_events.anomalous_download_alerts-0.1.0 is loaded   │ PASS   │         66ns │
│ box_events │ events                      │ asset     │ index_template logs-box_events.events is loaded                             │ PASS   │        205ns │
│ box_events │ events                      │ asset     │ ingest_pipeline logs-box_events.events-0.1.0 is loaded                      │ PASS   │        169ns │
│ box_events │ malicious_content_alerts    │ asset     │ index_template logs-box_events.malicious_content_alerts is loaded           │ PASS   │        196ns │
│ box_events │ malicious_content_alerts    │ asset     │ ingest_pipeline logs-box_events.malicious_content_alerts-0.1.0 is loaded    │ PASS   │        119ns │
│ box_events │ suspicious_locations_alerts │ asset     │ index_template logs-box_events.suspicious_locations_alerts is loaded        │ PASS   │        141ns │
│ box_events │ suspicious_locations_alerts │ asset     │ ingest_pipeline logs-box_events.suspicious_locations_alerts-0.1.0 is loaded │ PASS   │        185ns │
│ box_events │ suspicious_sessions_alerts  │ asset     │ index_template logs-box_events.suspicious_sessions_alerts is loaded         │ PASS   │        157ns │
│ box_events │ suspicious_sessions_alerts  │ asset     │ ingest_pipeline logs-box_events.suspicious_sessions_alerts-0.1.0 is loaded  │ PASS   │        108ns │
╰────────────┴─────────────────────────────┴───────────┴─────────────────────────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: box_events - END   ---
Done
Run pipeline tests for the package
--- Test results for package: box_events - START ---
╭────────────┬─────────────────────────────┬───────────┬───────────────────────────────┬────────┬──────────────╮
│ PACKAGE    │ DATA STREAM                 │ TEST TYPE │ TEST NAME                     │ RESULT │ TIME ELAPSED │
├────────────┼─────────────────────────────┼───────────┼───────────────────────────────┼────────┼──────────────┤
│ box_events │ anomalous_download_alerts   │ pipeline  │ test-anomalous-download.log   │ PASS   │   6.899616ms │
│ box_events │ events                      │ pipeline  │ test-copy.log                 │ PASS   │     2.5308ms │
│ box_events │ events                      │ pipeline  │ test-create.log               │ PASS   │    3.65869ms │
│ box_events │ events                      │ pipeline  │ test-download.log             │ PASS   │   3.548946ms │
│ box_events │ events                      │ pipeline  │ test-event-types.log          │ PASS   │  65.519602ms │
│ box_events │ events                      │ pipeline  │ test-preview.log              │ PASS   │   2.882847ms │
│ box_events │ events                      │ pipeline  │ test-rename.log               │ PASS   │   2.556427ms │
│ box_events │ events                      │ pipeline  │ test-trash.log                │ PASS   │   3.062823ms │
│ box_events │ events                      │ pipeline  │ test-upload.log               │ PASS   │   2.883961ms │
│ box_events │ malicious_content_alerts    │ pipeline  │ test-malicious-content.log    │ PASS   │   3.108011ms │
│ box_events │ suspicious_locations_alerts │ pipeline  │ test-suspicious-locations.log │ PASS   │   2.560171ms │
│ box_events │ suspicious_sessions_alerts  │ pipeline  │ test-suspicious-sessions.log  │ PASS   │    3.06026ms │
╰────────────┴─────────────────────────────┴───────────┴───────────────────────────────┴────────┴──────────────╯
--- Test results for package: box_events - END   ---
Done
Run static tests for the package
--- Test results for package: box_events - START ---
╭────────────┬─────────────────────────────┬───────────┬──────────────────────────┬────────┬──────────────╮
│ PACKAGE    │ DATA STREAM                 │ TEST TYPE │ TEST NAME                │ RESULT │ TIME ELAPSED │
├────────────┼─────────────────────────────┼───────────┼──────────────────────────┼────────┼──────────────┤
│ box_events │ anomalous_download_alerts   │ static    │ Verify sample_event.json │ PASS   │  44.056967ms │
│ box_events │ events                      │ static    │ Verify sample_event.json │ PASS   │  41.476353ms │
│ box_events │ malicious_content_alerts    │ static    │ Verify sample_event.json │ PASS   │  43.201784ms │
│ box_events │ suspicious_locations_alerts │ static    │ Verify sample_event.json │ PASS   │  40.801869ms │
│ box_events │ suspicious_sessions_alerts  │ static    │ Verify sample_event.json │ PASS   │  54.500493ms │
╰────────────┴─────────────────────────────┴───────────┴──────────────────────────┴────────┴──────────────╯
--- Test results for package: box_events - END   ---
Done
Run system tests for the package
--- Test results for package: box_events - START ---
No test results
--- Test results for package: box_events - END   ---
Done

elasticmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3677/detail/PR-3677/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3677/detail/PR-3677/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3677/detail/PR-3677/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3677/detail/PR-3677/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_3677.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-07-11T11:51:08.520Z&rangeTo=2022-07-11T12:11:08.520Z&transactionName=Ingest-manager/integrations/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=f409bb0cf025cb0fcb2469cb086966e2&transactionId=3d94819deedfacc4)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-07-11T12:01:08.520+0000


* Duration: 15 min 46 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 1  |
| Passed       | 0  |
| Skipped      | 0 |
| Total        | 1   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3677/detail/PR-3677/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `static test: Verify sample_event.json – box_events.events`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 null 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 one or more errors found in document: [0] field "source.parent.id" is undefined
[1] field "source.parent.type" is undefined
[2] field "source.parent.name" is undefined
[3] field "source.synced" is undefined
[4] field "source.modified_by.name" is undefined
[5] field "source.modified_by.id" is undefined
[6] field "source.modified_by.login" is undefined
[7] field "source.modified_by.type" is undefined
[8] field "source.content_modified_at" is undefined
[9] field "source.etag" is undefined
[10] field "source.id" is undefined
[11] field "source.item_status" is undefined
[12] field "source.created_by.type" is undefined
[13] field "source.created_by.name" is undefined
[14] field "source.created_by.id" is undefined
[15] field "source.created_by.login" is undefined
[16] field "source.path_collection.total_count" is undefined
[17] field "source.path_collection.entries" is undefined
[18] field "source.content_created_at" is undefined
[19] field "source.owned_by.name" is undefined
[20] field "source.owned_by.id" is undefined
[21] field "source.owned_by.login" is undefined
[22] field "source.owned_by.type" is undefined
[23] parsing field value failed: field "event.sequence"'s Go type, string, does not match the expected field type: long (field value: 0)
[24] field "recorded_at" is undefined
[25] field "event_id" is undefined
[26] field "session_id" is undefined 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3677/detail/PR-3677/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Test integration: box_events`
<ul>
<li>Took 0 min 4 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-3677/runs/1/steps/882/log/?start=0">here</a></li>
<li>Description: <code>                                eval "$(../../build/elastic-package stack shellinit)"                                ../../build/elastic-package test -v --report-format xUnit --report-output file --test-coverage                                </code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

djptek commented:
Pipeline tests and manual validation all OK

System tests WIP

Moved to Ready for Review to solicit comment in the interim

mtojek commented on packages/box_events/changelog.yml:
```diff
@@ -0,0 +1,6 @@
+  changes:
+    - description: Initial draft of the package
+      type: enhancement
+      link: https://github.com/elastic/integrations/pull/1 # FIXME Replace with the real PR link
```
Should be fixed?

mtojek reviewed (commented):
Left a few basic comments, didn't look deeper into this. Do you think that `box_events` is the correct name? Don't we want something more generic like `box`?

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

marc-gr commented on packages/box_events/data_stream/events/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,90 @@
+      ignore_missing: true
+  - rename:
+      field: box.created_by.type
+      target_field: event.type
```
Same as event.kind

djptek replied:
This value is always `user` see: https://developer.box.com/reference/resources/user--mini/#param-type which matches the accepted ECS value `user` https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-type.html#ecs-event-type-user

marc-gr commented on packages/box_events/data_stream/events/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,90 @@
+  - remove:
+      field: box.created_at
+      ignore_missing: true
+  - script:
```
Somewhere in the pipeline we would need to check for the `preserve_original_event` tag and act accordingly by removing `event.original` if not. eg https://github.com/elastic/integrations/blob/main/packages/google_workspace/data_stream/groups/elasticsearch/ingest_pipeline/default.yml#L292

djptek replied:
Agreed, I've added that into the pipeline 👍

marc-gr commented on packages/box_events/data_stream/events/fields/base-fields.yml:
```diff
@@ -0,0 +1,20 @@
+- name: event.module
+  type: constant_keyword
+  description: Event module
+  value: box
```
```suggestion
  value: box_events
```

djptek replied:
See https://github.com/elastic/integrations/pull/3677#issuecomment-1190238352 . I would have liked to call this `Box`, but there is already a `Box` connector from Workplace Search so I called it Box events to avoid ambiguity

marc-gr commented on packages/box_events/data_stream/events/fields/base-fields.yml:
```diff
@@ -0,0 +1,20 @@
+  description: Data stream type.
+- name: data_stream.dataset
+  type: constant_keyword
+  description: Data stream dataset.
```
not sure if the values for `data_stream.dataset` needs to be set or is automatically populated.

djptek commented on packages/box_events/manifest.yml:
```diff
@@ -0,0 +1,110 @@
+  - datastore
+release: experimental
+conditions:
+  kibana.version: "^8.2.0"
```
I've changed  `_dev/build` to 8.2.0 as that's what I'm testing against

djptek commented:
> Left a few basic comments, didn't look deeper into this. Do you think that `box_events` is the correct name? Don't we want something more generic like `box`?

@mtojek @marc-gr I would have liked to call this `Box`, but there is already a `Box` connector from Workplace Search

<img width="654" alt="image" src="https://user-images.githubusercontent.com/11043991/179984237-d0b31ba0-11d8-4293-8171-5ac478202fcf.png">

djptek commented on packages/box_events/_dev/build/docs/README.md:
```diff
@@ -0,0 +1,13 @@
+# Box Integration
+
+This integration periodically fetches events from [Box](https://app.box.com/). It can parse events created by Box on behalf of the user or enterprise. 
+
```
I added the config hints in the "Configure Integration" Pane here:

<img width="648" alt="image" src="https://user-images.githubusercontent.com/11043991/179985136-a5e6496c-82ac-4d68-a500-9bb94eb39239.png">

The issue with the credentials was resolved with the addition of `endpoint_params.*` to the Filebeat/Agent configuration. These properties are exposed in the "Configure Integration" Pane with extended comments describing how to navigate to find their respective values in the Box Api

djptek replied:
>> addition of `endpoint_params.*`

Thanks for that ☝️ @marc-gr

djptek replied:
@marc-gr I've added a lot more detail [here](https://github.com/elastic/integrations/blob/7817bee409c158dd2b19f129afbe677789175192/packages/box_events/_dev/build/docs/README.md#box-events) on authorisation and privileges

djptek commented on packages/box_events/data_stream/events/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,90 @@
+        - yyyy-MM-dd'T'HH:mm:ss-Z
+  - rename:
+      field: box.type
+      target_field: event.kind
```
Which matches `event` see: https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-kind.html#ecs-event-kind-event

djptek commented:
> I don't see it in elastic/integrations: https://github.com/elastic/integrations/search?q=%22search+over+your+files%22. Is it a Kibana plugin?

Yes, see : https://github.com/elastic/kibana/blob/0c78f8a439f082ef5dac5629e016d97279bd2711/x-pack/plugins/enterprise_search/server/integrations.ts#L30

djptek commented:
I updated the integrations splash page tile to better differentiate from Workplace Search Box Integration

<img width="639" alt="image" src="https://user-images.githubusercontent.com/11043991/180173239-3c59df40-b995-4fa5-ad44-f9b761160566.png">

djptek commented:
System Test Passed

```
2022/07/22 12:27:38 DEBUG deleting data in data stream...
--- Test results for package: box_events - START ---
╭────────────┬─────────────┬───────────┬────────────┬────────┬─────────────────╮
│ PACKAGE    │ DATA STREAM │ TEST TYPE │ TEST NAME  │ RESULT │    TIME ELAPSED │
├────────────┼─────────────┼───────────┼────────────┼────────┼─────────────────┤
│ box_events │ events      │ system    │ transforms │ PASS   │ 1m15.122175301s │
│ box_events │ events      │ system    │ oauth      │ PASS   │  2m6.521296743s │
╰────────────┴─────────────┴───────────┴────────────┴────────┴─────────────────╯
--- Test results for package: box_events - END   ---
Done

```

djptek commented:
Skipping the Oauth tests, asset & static tests return PASS

```
% elastic-package test -v                                                                          (box)integrations
2022/07/22 13:34:30 DEBUG Enable verbose logging
2022/07/22 13:34:31  INFO New version is available - v0.58.0. Download from: https://github.com/elastic/elastic-package/releases/tag/v0.58.0
Run test suite for the package
Run system tests for the package
--- Test results for package: box_events - START ---
No test results
--- Test results for package: box_events - END   ---
Done
Run asset tests for the package
2022/07/22 13:34:31 DEBUG installing package...
2022/07/22 13:34:31 DEBUG POST https://127.0.0.1:5601/api/fleet/epm/packages/box_events-0.1.0
--- Test results for package: box_events - START ---
╭────────────┬─────────────┬───────────┬─────────────────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE    │ DATA STREAM │ TEST TYPE │ TEST NAME                                                           │ RESULT │ TIME ELAPSED │
├────────────┼─────────────┼───────────┼─────────────────────────────────────────────────────────────────────┼────────┼──────────────┤
│ box_events │             │ asset     │ dashboard box_events-8883bd20-0770-11ed-b183-a74e67b2e31e is loaded │ PASS   │        610ns │
│ box_events │ events      │ asset     │ index_template logs-box_events.events is loaded                     │ PASS   │        139ns │
│ box_events │ events      │ asset     │ ingest_pipeline logs-box_events.events-0.1.0 is loaded              │ PASS   │        156ns │
╰────────────┴─────────────┴───────────┴─────────────────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: box_events - END   ---
Done
Run pipeline tests for the package
--- Test results for package: box_events - START ---
No test results
--- Test results for package: box_events - END   ---
Done
Run static tests for the package
--- Test results for package: box_events - START ---
╭────────────┬─────────────┬───────────┬──────────────────────────┬────────┬──────────────╮
│ PACKAGE    │ DATA STREAM │ TEST TYPE │ TEST NAME                │ RESULT │ TIME ELAPSED │
├────────────┼─────────────┼───────────┼──────────────────────────┼────────┼──────────────┤
│ box_events │ events      │ static    │ Verify sample_event.json │ PASS   │  47.590163ms │
╰────────────┴─────────────┴───────────┴──────────────────────────┴────────┴──────────────╯
--- Test results for package: box_events - END   ---
Done


```

djptek commented:
Added pipeline tests

```
% elastic-package test                                                                                                  (box)integrations
2022/07/22 14:29:06  INFO New version is available - v0.58.0. Download from: https://github.com/elastic/elastic-package/releases/tag/v0.58.0
Run test suite for the package
Run asset tests for the package
--- Test results for package: box_events - START ---
╭────────────┬─────────────┬───────────┬─────────────────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE    │ DATA STREAM │ TEST TYPE │ TEST NAME                                                           │ RESULT │ TIME ELAPSED │
├────────────┼─────────────┼───────────┼─────────────────────────────────────────────────────────────────────┼────────┼──────────────┤
│ box_events │             │ asset     │ dashboard box_events-8883bd20-0770-11ed-b183-a74e67b2e31e is loaded │ PASS   │      2.762µs │
│ box_events │ events      │ asset     │ index_template logs-box_events.events is loaded                     │ PASS   │        153ns │
│ box_events │ events      │ asset     │ ingest_pipeline logs-box_events.events-0.1.0 is loaded              │ PASS   │        172ns │
╰────────────┴─────────────┴───────────┴─────────────────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: box_events - END   ---
Done
Run pipeline tests for the package
--- Test results for package: box_events - START ---
╭────────────┬─────────────┬───────────┬───────────────────┬────────┬──────────────╮
│ PACKAGE    │ DATA STREAM │ TEST TYPE │ TEST NAME         │ RESULT │ TIME ELAPSED │
├────────────┼─────────────┼───────────┼───────────────────┼────────┼──────────────┤
│ box_events │ events      │ pipeline  │ test-copy.log     │ PASS   │   2.024754ms │
│ box_events │ events      │ pipeline  │ test-create.log   │ PASS   │   2.502694ms │
│ box_events │ events      │ pipeline  │ test-download.log │ PASS   │   3.008958ms │
│ box_events │ events      │ pipeline  │ test-preview.log  │ PASS   │   2.569494ms │
│ box_events │ events      │ pipeline  │ test-rename.log   │ PASS   │   2.206694ms │
│ box_events │ events      │ pipeline  │ test-trash.log    │ PASS   │   3.746997ms │
│ box_events │ events      │ pipeline  │ test-upload.log   │ PASS   │   2.699397ms │
╰────────────┴─────────────┴───────────┴───────────────────┴────────┴──────────────╯
--- Test results for package: box_events - END   ---
Done
Run static tests for the package
--- Test results for package: box_events - START ---
╭────────────┬─────────────┬───────────┬──────────────────────────┬────────┬──────────────╮
│ PACKAGE    │ DATA STREAM │ TEST TYPE │ TEST NAME                │ RESULT │ TIME ELAPSED │
├────────────┼─────────────┼───────────┼──────────────────────────┼────────┼──────────────┤
│ box_events │ events      │ static    │ Verify sample_event.json │ PASS   │  42.688147ms │
╰────────────┴─────────────┴───────────┴──────────────────────────┴────────┴──────────────╯
--- Test results for package: box_events - END   ---
Done
Run system tests for the package
--- Test results for package: box_events - START ---
No test results
--- Test results for package: box_events - END   ---
Done

```

djptek commented:
Re-requesting review after addressing comments and addition of pipeline tests

mtojek commented on packages/box_events/manifest.yml:
```diff
@@ -0,0 +1,110 @@
+categories:
+  - cloud
+  - datastore
+release: experimental
```
Release is [deprecated](https://github.com/elastic/package-spec/blob/main/versions/1/integration/manifest.spec.yml#L208) now.

djptek replied:
thanks @mtojek I've bumped the `version` to `1.0.0-beta` and removed `release`

mtojek reviewed (approved):
CODEOWNERS looks good.

jamiehynds commented:
> I updated the integrations splash page tile to better differentiate from Workplace Search Box Integration
> 
> <img alt="image" width="639" src="https://user-images.githubusercontent.com/11043991/180173239-3c59df40-b995-4fa5-ad44-f9b761160566.png">

@djptek to ensure consistency with our other integrations, can you please adjust the integration description to: `Collect logs from Box with Elastic Agent`

djptek commented:
@jamiehynds done

jamiehynds commented:
@djptek dashboard looks great, but some suggested modifications. 

- Possibly remove the Events by File size chart as I'm not sure that it adds much value. 
- User Activity: Table showing Top 10 Users x Number of Events associated with the user
- Top Users by Failed Login: Top 10 Users x number of Failed Logins
- Authentication by Country: Map showing geo location of login events, with a filter for Successful and Failed

jamiehynds commented:
@djptek can you confirm if [Shield](https://developer.box.com/guides/events/event-triggers/shield-alert-events/) events are covered by this integration or just the [Event Source ](https://developer.box.com/guides/events/event-triggers/event-source/)events?

I'm assuming Shield isn't covered, which is fine. Just want to ensure that I know what's covered for field enablement slides.

djptek commented:
>> can you confirm if [Shield](https://developer.box.com/guides/events/event-triggers/shield-alert-events/) events are covered by this integration or just the [Event Source ](https://developer.box.com/guides/events/event-triggers/event-source/)events?

@jamiehynds theoretically, yes, as the API call consumes the stream where these are produced

*All Shield threat detection alert events are produced within the [enterprise event](https://developer.box.com/guides/events/enterprise-events/for-enterprise//) stream.*

I would need a *Box Shield* license, or access to a suitable account in order to validate this.

*[Box Shield](https://www.box.com/shield) must be purchased and enabled on a Box enterprise in order to take advantage of the advanced security offerings outlined below.*

djptek commented:
@jamiehynds updated screenshot - no map I'm afraid, because no IP to run geo analysis on

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/11043991/180803671-d2fdeba4-566a-4dc7-b7c2-a4cb273678db.png">

djptek commented:
Updated screenshot after building dash vs 7.17
<img width="1518" alt="box_screenshot" src="https://user-images.githubusercontent.com/11043991/180987868-a6af18d2-d3b3-487e-a799-c0b535e30d7f.png">

djptek commented:
Back to Draft Pending addition of Box Shield Events

djptek commented:
@jamiehynds 

For BOX SHEILD, I'm adding the Threat Detection related events - context-rich alerts on potential threats, such as compromised accounts and data theft, based on anomalous user behavior.

[Threat detection alerts](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#threat-detection-alerts)
- [Suspicious locations alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#suspicious-locations-alert)
- [Suspicious sessions alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#suspicious-sessions-alert)
- [Anomalous download alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#anomalous-download-alert)
- [Malicious content alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#malicious-content-alert)

Do you think we should also add Smart Access Events - classification-based access policies to control access and prevent the unintentional leakage of sensitive content - or is that out-of-scope/do-later?

[Smart access](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#smart-access)
- [Download and Print Restriction](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#download-and-print-restriction)
- [External collaboration restriction](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#external-collaboration-restriction)
- [Application Restriction](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#application-restriction)
- [FTP Restriction](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#ftp-restriction)

djptek commented:
I'm breaking this down into multiple data streams, because the following two alerts require divergent `response.split` logic

[Suspicious locations alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#suspicious-locations-alert)
[Suspicious sessions alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#suspicious-sessions-alert)

`response.split` can respect the multiple levels of split required for both mappings, however, not at the same time, in the same data_stream. The workaround I'm using is to unpack the `json` in `httpjson.yml.json` in each data stream then `drop` any events whose mapping is not a match to that stream.

Theoretically, I could have just two data streams, e.g. one for locations and one for everything else. This would save a little client-side processing and API calls to Box, however, if there were to be changes to `response.split` in the future that might generate duplicate events, so I'm setting up 5 streams using the `drop` logic, one each for :

[Get events](https://developer.box.com/reference/get-events/)
[Suspicious locations alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#suspicious-locations-alert)
[Suspicious sessions alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#suspicious-sessions-alert)
[Anomalous download alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#anomalous-download-alert)
[Malicious content alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#malicious-content-alert)

djptek commented:
Added 4 x additional data streams for BOX SHIELD Alerts, with corresponding dashboard. Resubmitting for review

efd6 commented on packages/box_events/data_stream/malicious_content_alerts/_dev/test/pipeline/test-malicious-content.log-expected.json:
```diff
@@ -0,0 +1,134 @@
+                "kind": "alert",
+                "original": "{\"source\":null,\"created_by\":{\"type\":\"user\",\"id\":\"2\",\"name\":\"Unknown User\",\"login\":\"\"},\"action_by\":null,\"created_at\":\"2019-12-20T11:38:56-08:00\",\"event_id\":\"97f1b31f-f143-4777-81f8-1b557b39ca33\",\"event_type\":\"SHIELD_ALERT\",\"ip_address\":\"10.1.2.3\",\"type\":\"event\",\"session_id\":null,\"additional_details\":{\"shield_alert\":{\"rul ...
+                "risk_score": 100,
+                "type": "user"
```
"[indicator](https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-type.html#ecs-event-type-indicator)"?

djptek replied:
Now showing as:

```
            "event": {
                "action": "SHIELD_ALERT",
                "category": [
                    "threat"
                ],
                "id": "97f1b31f-f143-4777-81f8-1b557b39ca33",
                "kind": "alert",
                "risk_score": 77,
                "type": [
                    "access",
                    "indicator"
                ]
            }
```

efd6 commented on packages/box_events/data_stream/suspicious_locations_alerts/_dev/test/pipeline/test-common-config.yml:
```diff
@@ -0,0 +1,5 @@
+dynamic_fields:
+  event.ingested: ".*"
```
Remove.

djptek replied:
done

efd6 commented on packages/box_events/data_stream/suspicious_locations_alerts/_dev/test/pipeline/test-suspicious-locations.log-expected.json:
```diff
@@ -0,0 +1,95 @@
+                "kind": "alert",
+                "original": "{\"action_by\":null,\"additional_details\":{\"shield_alert\":{\"alert_id\":2398,\"alert_summary\":{\"alert_activities\":{\"event_type\":\"Download\",\"ip_info\":{\"city_name\":\"San Jose\",\"country_code\":\"US\",\"ip\":\"67.43.156.0\",\"latitude\":\"37.5555\",\"longitude\":\"-120.6789\",\"region_name\":\"California\",\"registrant\":\"Microsoft Corporation\"},\"item_i ...
+                "risk_score": 60,
+                "type": "user"
```
"indicator"?

djptek replied:
fixed, now showing:

```
            "event": {
                "action": "SHIELD_ALERT",
                "category": [
                    "network"
                ],
                "id": "97f1b31f-f143-4777-81f8-1b557b39ca33",
...
                "kind": "alert",
                "risk_score": 60,
                "type": [
                    "access",
                    "connection",
                    "indicator"
                ]
            }
```

efd6 commented on packages/box_events/data_stream/suspicious_sessions_alerts/_dev/test/pipeline/test-common-config.yml:
```diff
@@ -0,0 +1,5 @@
+dynamic_fields:
+  event.ingested: ".*"
```
Remove.

efd6 commented on packages/box_events/data_stream/suspicious_sessions_alerts/_dev/test/pipeline/test-suspicious-sessions.log-expected.json:
```diff
@@ -0,0 +1,187 @@
+                "kind": "alert",
+                "original": "{\"action_by\":null,\"additional_details\":{\"shield_alert\":{\"alert_id\":500,\"alert_summary\":{\"description\":\"First time in prior month user connected from ip 81.2.69.142 First time user agent Some User Agent (Some UA 4.5.6) appeared for user within prior month Apparent distance 9580.0 km between events 59 seconds apart is faster than possible\",\"sessions\":{\" ...
+                "risk_score": 77,
+                "type": "user"
```
"indicator"?

djptek replied:
fixed, now showing:

```
            "event": {
                "action": "SHIELD_ALERT",
                "category": [
                    "network"
                ],
                "id": "97f1b31f-f143-4777-81f8-1b557b39ca33",
                "kind": "alert",
...
                "risk_score": 77,
                "type": [
                    "access",
                    "connection",
                    "indicator"
                ]
            }
```

efd6 commented on packages/box_events/data_stream/anomalous_download_alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,213 @@
+#  - convert:
+#      field: related.ip
+#      type: ip
+  - geoip: 
+      field: threat.indicator.ip
+      target_field: threat.indicator.geo
```
Do AS look-ups as well?
```
  - geoip:
      database_file: GeoLite2-ASN.mmdb
      field: threat.indicator.ip
      target_field: threat.indicator.as
      properties:
        - asn
        - organization_name
      ignore_missing: true
  - rename:
      field: threat.indicator.as.asn
      target_field: threat.indicator.as.number
      ignore_missing: true
  - rename:
      field: threat.indicator.as.organization_name
      target_field: threat.indicator.as.organization.name
      ignore_missing: true
```

djptek replied:
done

djptek replied:
Added for the 4 threat-related data streams

efd6 reviewed (commented):
Please delete .t.tmp.swp

djptek commented on packages/box_events/_dev/build/build.yml:
```diff
@@ -0,0 +1,3 @@
+dependencies:
+  ecs:
+    reference: git@v8.2.0
```
thanks, updated to 8.4

djptek commented on packages/box_events/changelog.yml:
```diff
@@ -0,0 +1,11 @@
+# newer versions go on top
+- version: "1.0.0-beta"
+  changes:
+    - description: use version tag to respect deprecation of release
```
so, I'd been grepping through already-existing packages for how to set the release to `beta` and had mistakenly used the legacy `release: beta` yaml property. When I corrected that by using a postfix release tag I updated the changelog accordingly

djptek commented on packages/box_events/data_stream/anomalous_download_alerts/_dev/test/pipeline/test-anomalous-download.log:
```diff
@@ -0,0 +1,9 @@
+{"source":null,"created_by":{"type":"user","id":"2","name":"Unknown User","login":""},"action_by":null,"created_at":"2019-12-20T11:38:56-08:00","event_id":"97f1b31f-f143-4777-81f8-1b557b39ca33","event_type":"SHIELD_ALERT","ip_address":"10.1.2.3","type":"event","session_id":null,"additional_details":{"shield_alert":{"rule_category":"Anomalous Download","rule_id":123,"rule_name":"Anomalous Download ...
```
Done, echoed here for convenience 

The simulated shield events were created by extracting the events from the [Box Shield Alerts documentation](https://developer.box.com/guides/events/event-triggers/shield-alert-events/) and publishing these via a web proxy. They were then ingested using Filebeat running 4 distinct configurations, each based on the relevant httpjson.yml.hbs for the relevant data stream, to ensure that the distinct response.split.split and drop_event pipeline processors were applied and the resultant data routed to a temporary Elastic Stack. The resultant JSON was then merged with events generated by the ingest from the event data stream, to append the Elastic Agent related properties.

djptek commented on packages/box_events/data_stream/anomalous_download_alerts/_dev/test/pipeline/test-anomalous-download.log-expected.json:
```diff
@@ -0,0 +1,880 @@
+                "kind": "alert",
+                "original": "{\"source\":null,\"created_by\":{\"type\":\"user\",\"id\":\"2\",\"name\":\"Unknown User\",\"login\":\"\"},\"action_by\":null,\"created_at\":\"2019-12-20T11:38:56-08:00\",\"event_id\":\"97f1b31f-f143-4777-81f8-1b557b39ca33\",\"event_type\":\"SHIELD_ALERT\",\"ip_address\":\"10.1.2.3\",\"type\":\"event\",\"session_id\":null,\"additional_details\":{\"shield_alert\":{\"rul ...
+                "risk_score": 77,
+                "type": "user"
```
That's the original event, so `"type": "user"` is in the response to the api. Box API provides a `"source"` top level element which is != to e.g. an Elasticsearch `"_source"` element.

Even so, I had to look at the [developer docs]([Anomalous download alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#anomalous-download-alert)) and [support articles](https://support.box.com/hc/en-us/articles/4408259490195-Box-Shield-enhances-anomalous-download-detection) for this one.

The anomaly here is not unauthorised access, that would be the `suspicious sessions` alert. This is an authorised user making dubious use of their regular privileges, for example: `"description":"Significant increase in download content week over week, 9200% (25.04 MB) more than last week 12 additional files downloaded week over week)"`

So I actually qualified this as `"threat.indicator.type": "file"`- @efd6 let me know if you agree/disagree and I'll update accordingly

efd6 replied:
The reasoning that I had was that it was a download, and this would be classed as a `event.type` of `access`, but I guess it could also be `indicator` since it's anomalous. It would be worth getting another opinion.

djptek replied:
I initially misunderstood you were referring to the embedded `type` property. All became clear when I saw your subsequent comment on `copy`. Updated in pipeline & retested

djptek commented on packages/box_events/data_stream/anomalous_download_alerts/_dev/test/pipeline/test-common-config.yml:
```diff
@@ -0,0 +1,5 @@
+dynamic_fields:
+  event.ingested: ".*"
```
Thanks for spotting, I've removed from all 5 data streams

djptek commented on packages/box_events/data_stream/anomalous_download_alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,213 @@
+processors:
+  - set:
+      field: ecs.version
+      value: "8.3.0"
```
Thanks for spotting, updated accordingly

efd6 replied:
Has this been changed or is it being intentionally held at 8.3.0?

djptek commented on packages/box_events/data_stream/anomalous_download_alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,213 @@
+      ignore_failure: true
+      formats:
+        - yyyy-MM-dd'T'hh:mm:ssXXX
+      if: "ctx?.box?.source != null"
```
Thanks for spotting, corrected and re-tested in all 5 pipelines

djptek commented on packages/box_events/data_stream/anomalous_download_alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,213 @@
+#        ctx.related.ip = setList;
+#  - convert:
+#      field: related.ip
+#      type: ip
```
I'd left that in pending the outcome of [ECS issue 2046](https://github.com/elastic/ecs/issues/2046) 

I've removed it now, thanks

djptek commented on packages/box_events/data_stream/events/_dev/test/pipeline/test-common-config.yml:
```diff
@@ -0,0 +1,5 @@
+dynamic_fields:
+  event.ingested: ".*"
```
removed

djptek commented on packages/box_events/data_stream/events/_dev/test/pipeline/test-copy.log-expected.json:
```diff
@@ -0,0 +1,86 @@
+                "kind": "event",
+                "original": "{\"created_at\":\"2022-07-20T02:00:54-07:00\",\"created_by\":{\"id\":\"19530772260\",\"login\":\"info@elastic.co\",\"name\":\"Elastic Integrations\",\"type\":\"user\"},\"event_id\":\"b941f182663a862f0125c256a5f0cad4a837092b\",\"event_type\":\"ITEM_COPY\",\"recorded_at\":\"2022-07-20T02:00:55-07:00\",\"session_id\":null,\"source\":{\"content_created_at\":\"2022-07-20T0 ...
+                "sequence": 0,
+                "type": "user"
```
The API documentation for [Box Event Types](https://developer.box.com/guides/events/enterprise-events/for-enterprise/#event-types) lists 123 distinct property values and is incomplete:  *This list is not exhaustive, so it is possible events appear that are not listed.* Then there is [another list here](https://developer.box.com/reference/resources/event/#param-event_type) with 129 types. 

I created the test cases based on the documented subset of more common event types, I'll add a lookup to encapsulate your suggestions for copy etc. :+1:

djptek replied:
Lookup implemented, thanks for the hint to use `map:`

djptek commented on packages/box_events/data_stream/malicious_content_alerts/_dev/test/pipeline/test-common-config.yml:
```diff
@@ -0,0 +1,5 @@
+dynamic_fields:
+  event.ingested: ".*"
```
done

efd6 commented on packages/box_events/data_stream/events/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -36,10 +36,409 @@ processors:
+      description: Set event.category + event.type
+      lang: painless
+      source: |
+        List categories = new ArrayList();
```
Maybe use a params map to map between the `box.event_type` and the ECS `event.type` and `event.category`.

Also the place-holder values will need to be replaced with the real ECS-allowed values.

djptek replied:
Ah heck, the notification had abbreviated that to the last sentence so I'd missed the sample code & spent time today trying to figure it out before coming to a very similar conclusion :facepalm: 

Thanks anyway, this is very similar to what I did so I guess I was working along the right lines :+1:

efd6 commented on packages/box_events/data_stream/suspicious_locations_alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -21,18 +21,18 @@ processors:
   - set:
-      field: event.kind
-      value: "alert"
+      field: event.category
```
This should be an array I think.

djptek replied:
yep, done across all of 'em now

efd6 commented on packages/box_events/data_stream/suspicious_sessions_alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -23,10 +23,15 @@ processors:
-      target_field: event.type
-      ignore_missing: true
+  - set:
+      field: event.category
```
Array.

djptek replied:
done

efd6 commented on packages/box_events/data_stream/malicious_content_alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -28,18 +28,19 @@ processors:
-      field: event.kind
-      value: "alert"
+      field: event.category
+      value: "malware"
```
Array.

djptek replied:
done

efd6 commented on packages/box_events/data_stream/anomalous_download_alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -41,6 +41,11 @@ processors:
       ignore_failure: true
+  - set:
+      field: event.category
+      value: "threat"
```
Array.

djptek replied:
done

djptek commented on packages/box_events/data_stream/events/_dev/test/pipeline/test-download.log-expected.json:
```diff
@@ -0,0 +1,106 @@
+                "kind": "event",
+                "original": "{\"created_at\":\"2022-07-20T02:32:15-07:00\",\"created_by\":{\"id\":\"19530772260\",\"login\":\"info@elastic.co\",\"name\":\"Elastic Integrations\",\"type\":\"user\"},\"event_id\":\"0d836443a60c1e9cdf7f18ffa56d8b46b70a14c4\",\"event_type\":\"ITEM_DOWNLOAD\",\"recorded_at\":\"2022-07-20T02:32:16-07:00\",\"session_id\":\"lawelth4diwf8xuk\",\"source\":{\"content_created ...
+                "sequence": 0,
+                "type": "user"
```
Updated, now showing as:

```

{
    "event" : {
            "action": "ITEM_DOWNLOAD",
            "category": [
                   "file"
             ],
             "id": "0d836443a60c1e9cdf7f18ffa56d8b46b70a14c4",
             "kind": "event",
             "sequence": 0,
             "type": [
                   "access"
             ]
```

djptek commented on packages/box_events/data_stream/events/_dev/test/pipeline/test-preview.log-expected.json:
```diff
@@ -0,0 +1,106 @@
+                "kind": "event",
+                "original": "{\"created_at\":\"2022-07-19T07:18:13-07:00\",\"created_by\":{\"id\":\"19530772260\",\"login\":\"info@elastic.co\",\"name\":\"Elastic Integrations\",\"type\":\"user\"},\"event_id\":\"9fe7762deb9e4faea7af6ca5cfd2c8b6aee7fa8f\",\"event_type\":\"ITEM_PREVIEW\",\"recorded_at\":\"2022-07-19T07:18:14-07:00\",\"session_id\":\"j9fy4nh4cljnln47\",\"source\":{\"content_created_ ...
+                "sequence": 0,
+                "type": "user"
```
Updated, now showing as:

```

{
    "event" : {
            "action": "ITEM_DOWNLOAD",
            "category": [
                   "file"
             ],
...
             "kind": "event",
...
             "type": [
                   "access"
             ]
}
```

djptek commented on packages/box_events/data_stream/events/_dev/test/pipeline/test-rename.log-expected.json:
```diff
@@ -0,0 +1,86 @@
+                "kind": "event",
+                "original": "{\"created_at\":\"2022-07-20T02:01:29-07:00\",\"created_by\":{\"id\":\"19530772260\",\"login\":\"info@elastic.co\",\"name\":\"Elastic Integrations\",\"type\":\"user\"},\"event_id\":\"da27caec0e2d468df064f6baa7552b6e68ff4404\",\"event_type\":\"ITEM_RENAME\",\"recorded_at\":\"2022-07-20T02:01:30-07:00\",\"session_id\":null,\"source\":{\"content_created_at\":\"2022-07-20 ...
+                "sequence": 1,
+                "type": "user"
```
Updated, now showing as:

```

{
    "event" : {
            "action": "ITEM_DOWNLOAD",
            "category": [
                   "file",
                   "database"
             ],
...
             "kind": "change",
...
             "type": [
                   "access"
             ]
}
```

I added `database` where this reflects changes in the underlying document management system as well as the file storage blobs

djptek commented on packages/box_events/data_stream/events/_dev/test/pipeline/test-trash.log-expected.json:
```diff
@@ -0,0 +1,84 @@
+                "kind": "event",
+                "original": "{\"created_at\":\"2022-07-20T02:01:10-07:00\",\"created_by\":{\"id\":\"19530772260\",\"login\":\"info@elastic.co\",\"name\":\"Elastic Integrations\",\"type\":\"user\"},\"event_id\":\"8c1cfcf8fcb5a470cedcfd814bc333d6fac97837\",\"event_type\":\"ITEM_TRASH\",\"recorded_at\":\"2022-07-20T02:01:12-07:00\",\"session_id\":null,\"source\":{\"content_created_at\":\"2022-07-19T ...
+                "sequence": 1,
+                "type": "user"
```
Updated, now showing as:

```

{
    "event" : {
            "action": "ITEM_TRASH",
            "category": [
                   "file",
                   "database"
             ],
...
             "kind": "event",
...
             "type": [
                   "deletion"
             ]
}
```

djptek commented on packages/box_events/data_stream/events/_dev/test/pipeline/test-upload.log-expected.json:
```diff
@@ -0,0 +1,113 @@
+                "kind": "event",
+                "original": "{\"created_at\":\"2022-07-19T07:18:04-07:00\",\"created_by\":{\"id\":\"19530772260\",\"login\":\"info@elastic.co\",\"name\":\"Elastic Integrations\",\"type\":\"user\"},\"event_id\":\"36e2e8726b67f7ce275b479c4a0cb4ef0fd0d61c\",\"event_type\":\"ITEM_UPLOAD\",\"recorded_at\":\"2022-07-19T07:18:05-07:00\",\"session_id\":\"xp1dwzh4cljhofyq\",\"source\":{\"content_created_a ...
+                "sequence": 0,
+                "type": "user"
```
Fixed, now showing:

```  
            "event": {
                "action": "ITEM_UPLOAD",
                "category": [
                    "file"
                ],
                "id": "36e2e8726b67f7ce275b479c4a0cb4ef0fd0d61c",
                "kind": "event",
                "sequence": 0,
                "type": [
                    "creation"
                ]
            }
```

djptek commented:
Huge thanks for the extensive reviewing @efd6 

I've applied all your recommendations

efd6 commented on packages/box_events/data_stream/events/_dev/test/pipeline/test-event-types.log:
```diff
@@ -0,0 +1,130 @@
+{"event_type":"ACCESS_GRANTED"}
```
I like this test approach.

efd6 commented on packages/box_events/data_stream/anomalous_download_alerts/_dev/test/pipeline/test-anomalous-download.log-expected.json:
```diff
@@ -0,0 +1,967 @@
+            "event": {
+                "action": "SHIELD_ALERT",
+                "category": [
+                    "threat"
```
"file" for this as well?

djptek replied:
Added, now showing:

```
            "event": {
                "action": "SHIELD_ALERT",
                "category": [
                    "file",
                    "threat"
                ],
                "id": "97f1b31f-f143-4777-81f8-1b557b39ca33",
                "kind": "alert",
...
                "risk_score": 77,
                "type": [
                    "access",
                    "indicator"
                ]
            },

```

efd6 commented on packages/box_events/data_stream/events/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,724 @@
+            category: ['file','process']
+            type: ['deletion']
+      source: |
+        def setField(Map dest, String[] path, def value) {
```
Are you concerned about map sharing?

djptek replied:
Nope, I'd borrowed from elsewhere & agree it'd be neater to trim the code down as this is highly specific

Changed to

```
        def eventType = params.getOrDefault(ctx.box.event_type.toLowerCase(), null);
        if (eventType != null) {
          ctx.event.category = eventType.map.get('category');
          ctx.event.type = eventType.map.get('type');
        }
```

efd6 replied:
Thanks, that feel a lot better. If the keys of the params are left as the original then the work for the `toLowerCase` can be omitted as well.

djptek replied:
OK cool, I saw a lot of lowercase elsewhere & thought that was maybe a style thing. I've `awk`-ed them back to uppercase & removed the unneeded `.toLowerCase()`

djptek commented:
@efd6 re: 

```
      field: ecs.version
      value: "8.3.0"
```
>> Has this been changed or is it being intentionally held at 8.3.0?

Nope, I've been testing against `8.3` - makes more sense to move to `8.4` so I've updated across the board

efd6 reviewed (commented):
The pipeline test expectations need to be updated

It looks like the sample events were obtained from somewhere else (there are no system tests that would have generated them and the fields are not in lexical order as would be expected from generated samples). Is this the case

After the tests are fixed, this LGTM, but it would be good to get input on the event type mappings to ECS from someone else as well.

djptek commented:
@efd6 

>> It looks like the sample events were obtained from somewhere else (there are no system tests that would have generated them and the fields are not in lexical order as would be expected from generated samples). Is this the case

This is partially correct. 

- The event datastream is actually using real events extracted from the API during an earlier iteration of a system test, which required full setup of the API and wasn't practical to incorporate into the integration test.
- The pipeline tests for Shield events came from the original API documentation. I'll add an explanation, I think they better stay in Beta until there is a relevant license available

djptek commented:
[Root comment](https://github.com/elastic/integrations/pull/3677#issue-1300610798) updated with detail on origins of pipeline test events

djptek commented:
@kgeller @ebeahan @jamiehynds @epixa 

I'd be grateful if one of you might provide input on the event type mappings for the box integration, which are summarised in the following 5 sample documents, listed below with links to source documentation and target ingest document

# Box Events
[ ] original [Box Event](https://developer.box.com/reference/resources/event/) maps to Integration [events](https://github.com/djptek/integrations/tree/box/packages/box_events/data_stream/events/sample_event.json)

# Box Shield Alerts
[ ] original [Anomalous download alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#anomalous-download-alert) maps to Integration [anomalous_download_alerts](https://github.com/djptek/integrations/blob/box/packages/box_events/data_stream/anomalous_download_alerts/sample_event.json)
[ ] original [Malicious content alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#malicious-content-alert) maps to Integration [malicious_content_alerts](https://github.com/djptek/integrations/tree/box/packages/box_events/data_stream/malicious_content_alerts/sample_event.json)
[ ] original [Suspicious locations alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#suspicious-locations-alert) maps to Integration [suspicious_locations_alerts](https://github.com/djptek/integrations/tree/box/packages/box_events/data_stream/suspicious_locations_alerts/sample_event.json)
[ ] original [Suspicious sessions alert](https://developer.box.com/guides/events/event-triggers/shield-alert-events/#suspicious-sessions-alert) maps to Integration [suspicious_sessions_alerts](https://github.com/djptek/integrations/tree/box/packages/box_events/data_stream/suspicious_sessions_alerts/sample_event.json)

There is also an implicit mapping between the Box `event_type` field and ECS `event.category[]` and `event.type[]` arrays. The 3rd party documentation is brief, so I've made my best estimate for the 130 entries. If you want to take a look, the most concise representation is encapsulated in [this map](https://github.com/djptek/integrations/blob/8a4dd0d54acad3d43223b4594c9ce80f72ceb7ff/packages/box_events/data_stream/events/elasticsearch/ingest_pipeline/default.yml#L42)

kgeller commented:
@djptek LGTM! It aligns with my understanding, and what I've done for my two integrations so far

djptek commented:
Thanks @kgeller
